src/HTTP/Psr18Client.php

```diff
@@ -44,20 +46,50 @@ public function __construct(ClientInterface $httpClient, RequestFactoryInterface
      */
     public function request(string $method, string $url, array $headers = []): Response
     {
-        $uri = $this->uriFactory->createUri($url);
+        $permanentUrl = $url;
+        $requestedUrl = $url;
+        $remainingRedirects = $this->allowedRedirects;
 
-        $request = $this->requestFactory->createRequest($method, $uri);
+        $request = $this->requestFactory->createRequest(
+            $method,
+            $this->uriFactory->createUri($requestedUrl)
+        );
 
         foreach ($headers as $name => $value) {
             $request = $request->withHeader($name, $value);
         }
 
-        try {
-            $response = $this->httpClient->sendRequest($request);
-        } catch (ClientExceptionInterface $th) {
-            throw new HttpException($th->getMessage(), $th->getCode(), $th);
-        }
+        do {
+            $followRedirect = false;
+
+            try {
+                $response = $this->httpClient->sendRequest($request);
+            } catch (ClientExceptionInterface $th) {
+                throw new HttpException($th->getMessage(), $th->getCode(), $th);
+            }
+
+            $statusCode = $response->getStatusCode();
+
+            // If we have a redirect
+            if (in_array($statusCode, [300, 301, 302, 303, 307]) && $response->hasHeader('Location')) {
+                // Prevent infinity redirect loops
+                if ($remainingRedirects <= 0) {
+                    break;
+                }
+
+                $remainingRedirects--;
+                $followRedirect = true;
+
+                $requestedUrl = $response->getHeaderLine('Location');
+
+                if ($statusCode === 301) {
+                    $permanentUrl = $requestedUrl;
+                }
+
+                $request = $request->withUri($this->uriFactory->createUri($requestedUrl));
```

Comment:
Not sure what's the problem here.

Reply:
The problem is that the `location` header could have a non-absolute url (without host). But this is not a problem.

`$this->uriFactory->createUri()` creates a new Uri instance from the non-absolute url. This uri is set to `$request->withUri($uri)`. This will take care of the missing host:

> This method MUST update the Host header of the returned request by
>      * default if the URI contains a host component. **If the URI does not
>      * contain a host component, any pre-existing Host header MUST be carried
>      * over to the returned request.**

See https://github.com/php-fig/http-message/blob/2.0/src/RequestInterface.php#L99-L129